aodndata/aatams/aatams_sattag_nrt.py

```diff
@@ -0,0 +1,111 @@
+    return os.path.join(AATAMS_SATTAG_QC_NRT_BASE, os.path.basename(filepath))
+
+
+class AatamsSattagQcNRTHandler(HandlerBase):
+    """The class for aatams sattag qc NRT."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the class by only accepting zip files."""
+        super(AatamsSattagQcNRTHandler, self).__init__(*args, **kwargs)
+        self.allowed_extensions = [".zip"]
+        self.allowed_regexes = ["^.+_nrt\\.zip$"]
+        self.allowed_archive_path_regexes = [AATAMS_SATTAG_QC_NRT_BASE + ".+\\.zip"]
+        self.allowed_dest_path_regexes = [AATAMS_SATTAG_QC_NRT_BASE + ".+\\.csv"]
+        self.dest_path_function = dest_path
+        self.archive_input_file = True
+        self.archive_path_function = self.dest_path_function
+
+        # Initialize the schema class and set default validation method
+        self.schema = AatamsSattagQcDmSchema()  # Schema is the same as dm
+        self.validation_call = self.schema.extensive_validation
+        # Use below to just validate the file names and csv headers
+        # self.validation_call = self.SCHEMA.quick_validation
+        self.previous_files = None
+        self.broker = get_storage_broker(
+            self.config.pipeline_config["global"]["upload_uri"],
+        )
+
+    def is_update_required(self, old_metadata_file):
+        """Check if updates to the dataset/archive are required by inspecting the most
+        recent dates between the new stream and remote(current) metadata state."""
+        new_valid_data = self.schema.metadata
+        new_recent_date = sorted(new_valid_data["qc_end_date"])[-1]
+
+        _, old_valid_data = self.schema.validate_file(old_metadata_file)
+        old_recent_date = sorted(old_valid_data["qc_end_date"])[-1]
+        return new_recent_date > old_recent_date
+
+    def preprocess(self):
+        """Validate individual files within the zip."""
+        files_in_zip = self.file_collection.get_attribute_list("local_path")
+        self.validation_call(files_in_zip)
+
+        self.input_file_object.publish_type = PipelineFilePublishType.ARCHIVE_ONLY
+
+        self.file_collection.filter_by_attribute_value(
+            "file_type", FileType.CSV,
+        ).set_publish_types(PipelineFilePublishType.HARVEST_UPLOAD)
+
+    def process(self):
+        """Process NRT files, removing the previous files if content of new file is
+        updated."""
+
+        self.previous_files = self.state_query.query_storage(dest_path(""))
+        if self.previous_files:
+            old_metadata_file = self.previous_files[metadata_index(self.previous_files)]
+            new_metadata_file = self.file_collection[
+                metadata_index(self.file_collection)
+            ]
+            logger.info(
+                "NRT update requested: Comparing timestamps in the metadata files %s and %s"
+                % (old_metadata_file, new_metadata_file),
+            )
+            self.broker.download(old_metadata_file, self.temp_dir)
```

Comment:
In the past we've actively tried to avoid ever downloading the old files for comparison, and tried to compare metadata type things via either WFS (to access harvested metadata in the database) or OpeNDAP in the case of ACORN (which is less preferable). Any reason this can't similarly get metadata from harvested metadata?

In any case, it's definitely not the intention for handlers to make their own storage brokers. They are a low level library thing really.

If it absolutely needs to download a file instead of finding metadata from an existing source of metadata, we added a high level download capability to RemotePipelineFileCollection (which is the type returned by "state queries") for the products handler, e.g. https://github.com/aodn/python-aodndata/blob/master/aodndata/moorings/products_handler.py#L188

```python
from aodncore.pipeline.files import RemotePipelineFileCollection

old_metadata_file = self.previous_files[metadata_index(self.previous_files)]
collection_to_download = RemotePipelineFileCollection(old_metadata_file)
collection_to_download.download(old_metadata_file, self.temp_dir)
```

Reply:
Yeah, what @lwgordonimos said. Except...
> RemotePipelineFileCollection (which is the type returned by "state queries")

That was the intent, but I don't think that's been implemented yet. Looks like they just return a dict (e.g. [S3StorageBroker._run_query](https://github.com/aodn/python-aodncore/blob/841b1adf00f79f1378d99a9134413e4cce8641d8/aodncore/pipeline/storage.py#L352))

Also, the download method requires a broker, so the last line in your suggested snippet should be
```python
collection_to_download.download(self._upload_store_runner.broker, self.temp_dir)
```

Reply:
Wow, you're right. The old "intent versus reality" chestnut. I was sure I had changed that and it was all merged, but apparently only for the local filesystem broker...

I replicated that PR here for the S3 broker: https://github.com/aodn/python-aodncore/pull/208

Reply:
> Any reason this can't similarly get metadata from harvested metadata?

No reason - I just took the simple/least effort path for checking/testing using the available tools in the pipeline.
Implementing the WFS request would require more work/code/and is a bit more convoluted, particular in the tests.

I can certainly do it, but would wait until the pipeline requirements is more mature - the requirements may still change.

Reply:
Thanks for the heads up on the download thing - much better to use higher level functionality. 

Maybe it is a good idea to implemente the download method on the REmoteFile class - I was expecting that method to exist instead of using remotePipelineFileCollection.

Reply:
> Implementing the WFS request would require more work/code/and is a bit more convoluted, particular in the tests

If you eventually decide to do this, check out the [MooringsProductsHandler](https://github.com/aodn/python-aodndata/blob/master/aodndata/moorings/products_handler.py#L162), and [its tests](https://github.com/aodn/python-aodndata/blob/67ffb3038301aa104ba40970016e08ddb4bc69d9/test_aodndata/moorings/test_mooringsProductsHandler.py#L52) as examples

Reply:
FYI I have just promoted the changes in https://github.com/aodn/python-aodncore/pull/208 (v1.1.1) to prod, so the line above should now just be something like this
```suggestion
            self.download_remotepipelinefilecollection(old_metadata_file)
```

Reply:
yeap - I was just waiting for the changes to be avail. There are other lines too (like remove injection of self.broker namespace I think).